### ISSUE TITLE
firebase_hosting_custom_domain persistent DNS Records

### ISSUE THREAD

**Issue by TrieBr** [hashicorp/terraform-provider-google#16873]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### Description

TL;DR: Fix the API shape for supporting dynamic dns record creation, or add an example demonstrating how to do it in the current form.

`firebase_hosting_custom_domain` doesn't have a good way to dynamically create DNS records for the SSL challenges or other A/CNAME/TXT records. [required_dns_updates](https://registry.terraform.io/providers/hashicorp/google/5.5.0/docs/resources/firebase_hosting_custom_domain#required_dns_updates) is only present when the records need to be created so you can't reliably create DNS records from this field.

It would be nice is `required_dns_updates` was always populated regardless of if the DNS records have been previously verified or not, or even better: a simple field that is a list of [records](https://registry.terraform.io/providers/hashicorp/google/5.5.0/docs/resources/firebase_hosting_custom_domain#records) that should exist for the domain to be verified.

If this is already possible, it would be nice if the "Example Usage" included a section on how to dynamically create dns records based on this block (which I think would be an extremely common use case).

### New or Affected Resource(s)

* firebase_hosting_custom_domain

### Potential Terraform Configuration

```tf

resource "google_firebase_hosting_custom_domain" "default" {
  provider = google-beta
  project  = var.project_id
  site_id    = google_firebase_hosting_site.default.site_id
  custom_domain = var.dns_name
  wait_dns_verification = false
}
locals {
  # flatten ensures that this local value is a flat list of objects, rather
  # than a list of lists of objects.
  required_domain_records = flatten([
    for list_key, dns_updates in google_firebase_hosting_custom_domain.default.required_dns_updates : [
    for desired_key, desired in dns_updates.desired : [
      for record_key, record in desired.records : {
        list_key = list_key
        desired_key  = desired_key
        record_key  = record_key
        domain_name  = record.domain_name
        type  = record.type
        rdata  = record.rdata
      }
    ]
    ]
  ])
  depends_on = [google_firebase_hosting_custom_domain.default]
}

resource "google_dns_record_set" "domains" {
  for_each = local.required_domain_records
  project     = var.project_id
  name = "${each.value.domain_name}."
  type = each.value.type
  ttl  = 300
  managed_zone = var.root_dns_managed_zone_name
  rrdatas = [each.value.rdata]
  depends_on = [google_firebase_hosting_custom_domain.default]
}

```

The above example illustrates how dns records could be dynamically generated using the output from `google_firebase_hosting_custom_domain`, but this doesn't work as-is mainly because `required_dns_updates` can change between runs.

### References

* #0000

**Comment by melinath** [comment]
Note from triage: It sounds like the issue here is that google_firebase_hosting_custom_domain.required_dns_updates.desired only contains items the need to exist (but don't yet exist), which means that once they are created, the field is empty - so iterating over the field to automate domain creation will create them correctly but then try to delete them immediately afterward. Is that accurate?

**Comment by BenJackGill** [comment]
I am having the same issue. Struggling to figure out how to get the A/CNAME/TXT records that are required for verification.

@TrieBr Did you manage to overcome this issue?

**Comment by anuraaga** [comment]
I also ran into this issue, it will be great if it's possible to automate DNS verification for Firebase hosting. A couple of ideas for the proposal

- It might be just my inexperience with it, but I couldn't find a way to flatten two complex lists with Terraform CDK (e.g. typescript). Making sure the DNS verification records is a single list may make it easier to use in general, and especially with CDK

- I have a scheme where I delegate the nameservers of a subdomain, e.g. `alpha.project.com` to a separate GCP managed zone for `alpha.project.com`. I'm not sure if it's a GCP bug or not, but I'm unable to add a CNAME to such a zone since it is considered "not a subdomain". If it's possible to access the advanced verification records in addition to the quick setup records, it would be nice for this case.

**Comment by BenJackGill** [comment]
Any updates on this? Six months and no resolution?

**Comment by Ranguna** [comment]
It does make sense to me for the `discovered` and `desired` values to be dynamic, because they depend on which records already exist for a specific domain.

Maybe a new field could be added here next to these two, something like `required`. This new field will contain the dns record that are always required, regardless of the records that already exist.

Would that work ?

**Comment by Ranguna** [comment]
@anuraaga this worked for me in cdk (I added comment showing what I think the intermediate types are):

```
TerraformIterator.fromMap(Fn.flatten(
	TerraformIterator.fromList(
		Fn.flatten(
			TerraformIterator.fromList(
				firebaseHostingCustomDomain.requiredDnsUpdates
				/*
					list(object(
						desired = list(object(
							records=list(object(
								domain = string,
								type = string,
								rdata = string
							))
						))
					))
				*/
			).pluckProperty("desired")
			/*
				list(
					list(object(
						records=list(object(
							domain = string,
							type = string,
							rdata = string
						))
					))
				)
			*/
		)
		/*
			list(object(
				records=list(object(
					domain = string,
					type = string,
					rdata = string
				))
			))
		*/
	).pluckProperty("records")
	/*
		list(
			list(object(
				domain = string,
				type = string,
				rdata = string
			))
		)
	*/
));
/*
	list(
		object(
			domain = string,
			type = string,
			rdata = string
		)
	)
*/
```

But this does not solve the issue where the `desired` value is empty the second time I deploy, and thus, making the record resource disappear.

**Comment by anuraaga** [comment]
Thanks for the snippet! If going with the current structure, having that to refer to somewhere will be quite helpful. But it does seem like a beast so maybe it's worth simplifying if making a new field.

I think having a separate required property could help fix this issue. But tbh, I'm not sure if having a dynamic field like desired is even idiomatic Terraform, I think the general intention is for applying a resource multiple times to have no changes even to outputs, which isn't possible with the dynamic fields. Given that, I suspect the fields don't have any realistic use case anyways and should likely be removed or replaced.

**Comment by Ranguna** [comment]
Yep agreed. Terraform is supposed to be idempotent between runs, and this clearly is not.

**Comment by Ranguna** [comment]
Here's a possible workaround to this (although incredibly ugly imo):

Apply terraform twice, once to capture all the required dns records and store them in a cache, and twice to actually create the records.

1. `terraform apply`, terraform will:
   1. Create the all the necessary firebase resources, including the hosting site and custom domain;
   2. Unpack the records out of `required_domain_records` and `cert` complex lists;
   3. Store the result in json (e.g. into a bucket object in S3/gcp storage/somewhere else) by appending to what is already in cache.
2. `terraform apply` again, terraform will do everything from the step above, but additionally:
   1. Read the records from the cached json;
   2. Create the DNS records that were read from the cache.

The trick here is to append to the cache and not simply replace the contents in cache on every run. If we were to replace the contents of the cache, we would essentially have the same problem on the 4rd run, where:

1. 1st run would create the firebase custom domain and store the records in cache;
2. the 2nd run would create the records cached from the 1st run;
3. the firebase custom domain resource from 3rd run would report 0 dns records and terraform would write 0 records into the cache;
4. and finally, the 4rd run would read 0 records from the cache and remove the records from the DNS.

Here's a pseudo cdktf in ts code:

```ts
// utility function used below to unpack dns records from their deeply nested complex lists
const flattenDeeplyNestedListOfObjects = (record: Parameters<typeof TerraformIterator.fromList>[0], paths: string[]) =>
	paths.reduce((flattenList, path) =>
		Fn.flatten(
			TerraformIterator.fromList(
				flattenList
			).pluckProperty(path)
		),
		record
	);
...

const customDomain = new GoogleFirebaseHostingCustomDomain(this, 'firebase_hosting_custom_domain', {
	...
});

const dnsRecords = Fn.distinct(Fn.tolist(
	TerraformIterator.fromList(Fn.concat([
		flattenDeeplyNestedListOfObjects(customDomain.requiredDnsUpdates, [
			"desired",
			"records"
		]),
		flattenDeeplyNestedListOfObjects(customDomain.cert, [
			"verification",
			"dns",
			"desired",
			"records",
		])
	])).forExpressionForList(`{
		domain_name = val.domain_name
		type = val.type
		rdata = val.rdata
	}`))
);

const gcsBackend = new GcsBackend(this, { // GcsBackend or wherever you are storing your state
	...
});
const remoteBackendData = gcsBackend.getRemoteStateDataSource(this, 'remote_backend_data', YOUR_STACK_ID);

let dnsRecordCacheBucketObjectRead = new DataGoogleStorageBucketObjectContent(this, 'dns_record_cache_read', { // DataGoogleStorageBucketObjectContent or wherever you want to store the cache
	name: 'dns-record-cache.json',

	...

	// only read the cache if the stack has been deployed at least once
	// or else terraform will fail saying that this object does not exist
	count: Fn.try([Fn.conditional(Fn.lookupNested(remoteBackendData, ['outputs', 'deployed']), 1, 0), 0])
});

// map the dns records to a simpler object
const dnsRecordsToWriteToDns = Fn.jsondecode(Fn.try([Fn.lookupNested(dnsRecordCacheBucketObjectRead, [0, "content"]), '[]']));
const dnsRecordsToWriteToCache = Fn.distinct(Fn.concat([
	dnsRecords,
	dnsRecordsToWriteToDns
]));

new GoogleStorageBucketObject(this, 'dns_record_cache_write', { // GoogleStorageBucketObject or wherever you want to store you cache
	name: 'dns-record-cache.json',

	...

	content: Fn.jsonencode(dnsRecordsToWriteToCache),
});

const dnsRecordsToCreateIterator = TerraformIterator.fromMap(dnsRecordsToWriteToDns);

new Record(this, 'dns_record', {
        // if the record order were to change in the cache, terraform would destroy and recreate the same records
        // so we cover the array into a map by doing a base64 of this string just so the order of the records does not matter in the array
	forEach: TerraformIterator.fromMap(dnsRecordsToCreateIterator.forExpressionForMap('base64encode("${val.domain_name}-${val.type}")', "val")),
	
	name: dnsRecordsToCreateIterator.getString("domain_name"),
	type: dnsRecordsToCreateIterator.getString("type"),
	content: dnsRecordsToCreateIterator.getString("rdata"),

	dependsOn: [
		dnsRecordCacheBucketObjectRead,
		customDomain
	]
});
```

This code is comprised of snippets from my own code and it might need changes for it to fully work.

To give you some context, the above is the workaround I'm doing instead of simply:

```ts
...

const customDomain = new GoogleFirebaseHostingCustomDomain(this, `firebase_hosting_custom_domain`, {
	...
});

const dnsRecordsToCreateIterator = TerraformIterator.fromMap(customDomain.dns_records);

new Record(this, 'dns_record', {
	forEach: TerraformIterator.fromMap(dnsRecordsToCreateIterator),
	
	name: dnsRecordsToCreateIterator.getString("domain_name"),
	type: dnsRecordsToCreateIterator.getString("type"),
	content: dnsRecordsToCreateIterator.getString("rdata"),

	dependsOn: [customDomain]
})
```

If `GoogleFirebaseHostingCustomDomain` returned an idempotent map of the dns records that need to be created.

**Comment by Ranguna** [comment]
I took a deeper look at how this property is set here: https://github.com/hashicorp/terraform-provider-google-beta/blob/3c53a9ba48b510dd4e77724ebbb6aeb52c2783e0/google-beta/services/firebasehosting/resource_firebase_hosting_custom_domain.go#L717

And it seems like it's just a dumb proxy to an api call to firebase, which has the following response body: https://firebase.google.com/docs/reference/hosting/rest/v1beta1/projects.sites.customDomains

Meaning that this goes far beyond the code in this terraform provider. It seems like the Firebase api needs to be changed to accommodate this terraform module idempotency needs.

Welp, this makes me lose hope on a solution in the near future.

**Comment by BenJackGill** [comment]
This is my current workaround. It uses parts of the "advanced setup" method that the Firebase console gives as an option, and also the SSL verification records provided by the terraform module.

This is what I use to get a persistant SSL verification record:
```Terraform
# Add a custom domain to the Firebase Hosting site
resource "google_firebase_hosting_custom_domain" "main" {
  provider      = google-beta
  project       = var.gcp_project_id
  site_id       = google_firebase_hosting_site.main.site_id
  custom_domain = var.firebase_custom_domain
}

locals {
  # This will always return one record because we are using the "one" function below
  ssl_verification_dns_record = one(flatten([
    # Get the desired records - the verification record will come from here before it is added
    [
      for cert in google_firebase_hosting_custom_domain.main.cert : [
        for ver in cert.verification : [
          for dns in ver.dns : [
            for desired in dns.desired : [
              # Sometimes firebase will incorrectly have "REMOVE" as the required_action for the records.
              # We filter for "ADD" to avoid that.
              for record in desired.records : record if record.required_action == "ADD"
            ]
          ]
        ]
      ]
    ],
    # Get the discovered records - the verification record will come from here after it is added
    [
      for cert in google_firebase_hosting_custom_domain.main.cert : [
        for ver in cert.verification : [
          for dns in ver.dns : [
            for discovered in dns.discovered : [
              for record in discovered.records : record
            ]
          ]
        ]
      ]
    ],
  ]))
}
```

Then I use that in combination with some hardcoded values to verify the domain and point the domain to firebase. Use these dns records with whatever domain provider you have.
```Terraform
  dns_records = [
    # Firebase domain verification. Advanced setup. Proves ownership for domain.
    {
      name    = var.firebase_custom_domain # e.g. "app.example.com"
      type    = "TXT"
      content = "hosting-site=${var.gcp_project_id}" # e.g. "hosting-site=my-project-12345"
      proxied = false
    },
    # Firebase SSL verification Advanced setup. Proves ownership for SSL cert.
    {
      name    = module.firebase.ssl_verification_dns_record.domain_name # e.g. "_acme-challenge.app.example.com"
      content = module.firebase.ssl_verification_dns_record.rdata # e.g. "kjhasdfgq376fg738diugqwdukg"
      type    = module.firebase.ssl_verification_dns_record.type # e.g. "TXT"
      proxied = false
    },
    # Firebase custom domain hosting. Points the domain to the firebase project.
    {
      name    = var.firebase_custom_domain # e.g. "app.example.com"
      type    = "CNAME"
      content = "${var.gcp_project_id}.web.app" # e.g. "my-project-12345.web.app"
      proxied = true
    },
  ]
```

**Comment by schrodervictor** [comment]
It's April, 2026, and this issue is still here. The way the list of DNS records is populated and returned by this TF resource is not aligned with the idea of idempotency. What we need here is a persistent list of required DNS records for the hosting at Firebase to be valid with a custom domain, regardless if they were validated or not.
